### PR TITLE
Add Deep Video Deinterlacing Train & Test code

### PR DESCRIPTION
**Pull request by JoeyBallentine**
Added a custom PyTorch implementation of this paper: https://arxiv.org/pdf/1708.00187.pdf

Original TF code: https://github.com/HENDRIX-ZT2/Deep-Video-Deinterlacing

### CODE REVIEW

File: codes/dataops/augmentations.py [hunk 1]
```diff
@@ -502,6 +502,20 @@ def random_crop_pairs(img_HR, img_LR, HR_size, scale):
     
     return img_HR, img_LR
 
+
+def random_crop_dvd(img_top, img_bot, img_in, patch_size):
```
Comment (victorca25) [inline codes/dataops/augmentations.py:506]: Should reuse https://github.com/victorca25/BasicSR/blob/6b2317617c957e0805bcefbe614428e6f36fda00/codes/data/Vid_dataset.py#L487 . May be better to convert into a reusable class

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import torch
+import torch.nn as nn
+
+def vertcal_upscale(x, upfield=True):
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:4]: typo: vertical

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import torch
+import torch.nn as nn
+
+def vertcal_upscale(x, upfield=True):
+
+    n, c, h, w = x.shape
+    h *= 2
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:7]: Is there any possibility of h not being exactly twice the size?

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+def vertcal_upscale(x, upfield=True):
+
+    n, c, h, w = x.shape
+    h *= 2
+    t = x
+    if upfield:
+        out = torch.cat([t, torch.zeros_like(t)], 3)
+    else:
+        out = torch.cat([torch.zeros_like(t), t], 3)
+    out = torch.reshape(out, (n, c, h, w))
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:13]: Here could be safer to reshape to (n, c, -1, w) , the correct shape can be infered

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    return out
+
+
+def replace_field(x, input_image, upfield=True):
+
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:18]: Here might need a check if H is an even or odd value

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    upper_input = input_image[:, :, 0::2, :]
+    lower_input = input_image[:, :, 1::2, :]
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:20]: If H is an even value, would lower_input be missing one line?

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    # print(upper_input.shape, lower_input.shape)
+
+    if upfield:
+        # print(upper_input.shape, x.shape)
+        x = vertcal_upscale(x, upfield=False)
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:25]: typo? vertical

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        conv_fea_1 = nn.Sequential(nn.Conv2d(in_nc, nf, 3), nn.ReLU())
+        conv_fea_2 = nn.Sequential(nn.Conv2d(nf, nf, 3), nn.ReLU())
+        conv_fea_3 = nn.Conv2d(nf, nf//2, 1)
+        h = nn.Sequential(conv_fea_1, conv_fea_2, conv_fea_3)
+
+        conv_branch_top = nn.Conv2d(nf//2, nf//2, 3)
+        conv_branch_bottom = nn.Conv2d(nf//2, nf//2, 3)
+
+        final_branch_top = nn.Conv2d(
+            nf//2, out_nc, 3, stride=(2, 1), padding=4, padding_mode='replicate')
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:53]: Both H and W dimensions need padding? Else, use correct (H, W) tuple

File: codes/models/modules/architectures/DVDNet_arch.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        final_branch_bottom = nn.Conv2d(
+            nf//2, out_nc, 3, stride=(2, 1), padding=4, padding_mode='replicate')
```
Comment (victorca25) [inline codes/models/modules/architectures/DVDNet_arch.py:55]: Same padding comment

File: codes/options/options.py [hunk 10]
```diff
@@ -112,7 +112,7 @@ def dict_constructor(loader, node):
             opt = yaml.load(f, Loader=Loader)
 
     opt['is_train'] = is_train
-    scale = opt['scale']
+    scale = opt.get('scale', 1)
```
Comment (victorca25) [inline codes/options/options.py:115]: defatult to 1 or 4? Default ESRGAN is 4

Comment (rlaPHOENiX) [reply codes/options/options.py:115]: This wouldn't be used for upscaling, so a scale of anything other than 1 could be considered an error.
It's probably better to even remove the scale from the opt file and always use `1`.

Comment (rlaPHOENiX) [reply codes/options/options.py:115]: But it is true that this should probably default as 4 instead of 1, since this is also used for everything else. Perhaps before calling options.parse, force the scale to 1 in test_dvd?

Comment (JoeyBallentine) [reply codes/options/options.py:115]: The problem is I can't not include scale in the options file or this part of the code throws a fit since it doesn't have a default. 1 makes sense in my opinion as it would be the scale for any network that doesn't actually use scale, ie this, style transfer, etc. If there's another way to do this let me know

File: codes/models/modules/architectures/DVDNet_arch.py (hunk 3 above)
Comment (JoeyBallentine) [reply codes/models/modules/architectures/DVDNet_arch.py:7]: Because real interlaced data is made of fields that are both the same size, there's no way for adding both of those together to not equal exactly h*2. I suppose someone could input something that isn't like real world interlacing but tbh I don't really see that being an issue

Comment (rlaPHOENiX) [reply codes/models/modules/architectures/DVDNet_arch.py:7]: Yeah both fields HAVE to be the same height due to how it works, so this isn't a problem.

File: codes/models/modules/architectures/DVDNet_arch.py (hunk 5 above)
Comment (JoeyBallentine) [reply codes/models/modules/architectures/DVDNet_arch.py:18]: Similar to the other issue I do not think this will be a realistic issue as interlaced frames will always have an even height

File: codes/models/modules/architectures/DVDNet_arch.py (hunk 6 above)
Comment (JoeyBallentine) [reply codes/models/modules/architectures/DVDNet_arch.py:20]: Yes, but aimilar to the other issue I do not think this will be a realistic issue as interlaced frames will always have an even height

File: codes/models/modules/architectures/DVDNet_arch.py (hunk 8 above)
Comment (JoeyBallentine) [reply codes/models/modules/architectures/DVDNet_arch.py:53]: both need padding of 4 for some reason

**Comment by JoeyBallentine** [comment]
Ok I have now fixed everything that had an issue. I do think we need to investigate why the padding is necessary though.

**Comment by victorca25** [comment]
Add Deep Video Deinterlacing, still need to add reference paper in WIP in the Readme
